src/System.IdentityModel.Tokens/AsymmetricSignatureProvider.cs

```diff
@@ -83,41 +83,58 @@ public AsymmetricSignatureProvider(AsymmetricSecurityKey key, string algorithm,
                 throw new ArgumentOutOfRangeException("key.KeySize", key.KeySize, string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10630, key.GetType(), SignatureProviderFactory.MinimumAsymmetricKeySizeInBitsForVerifying));
             }
 
+#if DNXCORE50
+            hash = GetHashAlgorithmName(algorithm);
+#else
             hash = GetHashAlgorithm(algorithm);
+#endif
             RsaSecurityKey rsaKey = key as RsaSecurityKey;
+
             if (rsaKey != null)
             {
+#if DNXCORE50
+                rsaCng = new RSACng();
+                (rsaCng as RSA).ImportParameters(rsaKey.Parameters);
+#else
                 rsaCryptoServiceProvider = new RSACryptoServiceProvider();
                 (rsaCryptoServiceProvider as RSA).ImportParameters(rsaKey.Parameters);
+#endif
                 return;    
             }
 
             X509SecurityKey x509Key = key as X509SecurityKey;
             if (x509Key != null)
             {
-                RSACryptoServiceProvider rsa = null;
+
+#if DNXCORE50
                 if (willCreateSignatures)
                 {
-					rsa = x509Key.PrivateKey as RSACryptoServiceProvider;
+                    rsaCng = RSACertificateExtensions.GetRSAPrivateKey(x509Key.Certificate) as RSACng;
```

Comment:
Same remark. Will `GetRSAPrivateKey` ALWAYS return a CNG-based instance? I'm pretty sure I read somewhere there was a degraded mode falling back to CAPI when CNG is unavailable.